Fixed Issue 12752: std.algorithm.isPermutation

JackStouffer opened the pull request:
Add the template function isPermutation to determine if two ranges are permutations

https://issues.dlang.org/show_bug.cgi?id=12752

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,39 @@ unittest
+    import std.algorithm.searching : countUntil;
+
+    foreach (item; r1) {
+        if (r2.countUntil(item) == -1) {
```
This is incorrect, the sig constraints only specify that `r2` is an input range, but this code only works if `r2` is iterated from the beginning each time. That means it must be a forward range, and this should be checked in the sig constraints, and `save` should be explicitly called here to prevent exhausting the range after the first loop iteration.

quickfur replied:
Also, this algorithm is `O(n*m)` where `n` and `m` are the lengths of the respective ranges. Surely there are more efficient algorithms that achieve the same result?

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,39 @@ unittest
+    isInputRange!Range2 &&
+    !isInfinite!Range1 && 
+    !isInfinite!Range1
+) {
```
Style nitpick: please put the `)` at the end of the previous line.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,39 @@ unittest
+    assert(isPermutation([1, 2, 3], [3, 2, 1]));
+    assert(isPermutation([1.1, 2.3, 3.5], [2.3, 3.5, 1.1]));
+    assert(isPermutation("abc", "bca"));
+    assert(!isPermutation([1, 2], [3, 4]));
```
Need to test with non-array ranges as well. Maybe the stuff in `std.internal.test.dummyrange` may be useful in constructing such test ranges.

JackStouffer commented:
Thanks for the comments. I will fix these soon.

yebblies commented:
IIUC this will return `true` for `[1, 1, 2, 3]` and `[1, 2, 2, 3]` as each element in the first list is in the second.

quickfur commented:
Sorting the two ranges and comparing will yield a more efficient algorithm (O(m log n + n log m) rather than O(m*n)), as well as address the duplicated element issue raised by @yebblies .  However, it will require some extra work if the incoming ranges should not be mutated, and may need to allocate in that case.

JackStouffer commented:
That would require the ranges to be random access ranges. And in my current version, I have an overload for that case. But, I am still having trouble working on your other suggestion to get ReferenceForwardRange working with it.

yebblies commented:
An option that avoids needing O(n) storage is to _count_ the number of occurrences and check they match for each element.

It should also check for length mismatch is both ranges have length available.

I'm seeing three distinct uses here:
- In-place sorted - `equal(sort(r1), sort(r2))`
- Sorted copy - `equal(sort(r1.array), sort(r2.array))`
- O(1) space - the current one

quickfur commented:
An interesting thought occurred to me: if both ranges are input ranges (non-forward, meaning only single pass), you can implement a scoreboard algorithm. Create an AA mapping elements to integers, and for each element in the first range, increment the corresponding AA entry, and for each element in the second range, decrement the corresponding AA entry. At the end, all entries should be zero, otherwise the ranges were not permutations of each other. You can delete zero entries from the AA while iterating, and then you only have to check whether the AA is empty at the end.

This will require memory allocation, though. But at least it will be usable on input non-forward ranges.

JackStouffer commented:
I have this in my current code :)

```
enum AllocateGC {
    yes,
    no
};

bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2, const AllocateGC allocate_gc = AllocateGC.no) if (
    isForwardRange!Range1 &&
    isForwardRange!Range2 &&
    !isRandomAccessRange!Range1 &&
    !isRandomAccessRange!Range2 &&
    !isInfinite!Range1 && 
    !isInfinite!Range1)
{
    if (allocate_gc == AllocateGC.yes) {
        import std.conv: to;

        // If the two ranges aren't the same length, then they cannot
        // possibly be permutations
        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length")) {
            if (r1.length != r2.length) {
                return false;
            }
        }

        int[string] r1_counts;
        int[string] r2_counts;

        foreach (item; r1) {
            ++r1_counts[to!string(item)];
        }

        foreach (item; r2) {
            ++r2_counts[to!string(item)];
        }

        return r1_counts == r2_counts;
    } else {
        import std.algorithm.searching : count;

        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length")) {
            if (r1.length != r2.length) {
                return false;
            }
        }

        foreach (item; r1) {
            if (r1.count(item) != r2.count(item)) {
                return false;
            }
        }

        return true;
    }
}

bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2) if (
    isRandomAccessRange!Range1 &&
    isRandomAccessRange!Range2 &&
    !isInfinite!Range1 && 
    !isInfinite!Range1)
{
    import std.algorithm.sorting : sort;

    static if (hasMember!(Range1, "length") && hasMember!(Range2, "length")) {
        if (r1.length != r2.length) {
            return false;
        }
    }

    Range1 r1_copy = r1.save();
    Range2 r2_copy = r2.save();

    sort(r1_copy);
    sort(r2_copy);

    return r1_copy == r2_copy;
}
```

It's literally just the reference forward range that I am having trouble with.

quickfur commented:
Whoa. Your non-allocating forward range case has an O(m^2 * n^2) loop. We need to think about improving that, otherwise it's going to be a huge bottleneck on ranges of non-trivial length.

As for reference ranges, I don't see why it should be a problem. One thing that perhaps might be related, is that any time you need to iterate a range more than once, you should use `.save`, regardless of whether the range is reference or not. The reason is that the range could be a by-value wrapper around a by-reference data source that gets consumed once you iterate over it. Since you're requiring forward ranges, you should always use `.save` when iterating over the source ranges to preserve the correct semantics in all cases.

JackStouffer commented:
@quickfur Ok, I think I correctly addressed the reference range issue. Thanks for your advice.

Also, I'm not quite sure what to do about the non allocating forward range case. All of the permutation algorithms that I see sort the two inputs then compare them, which isn't possible for that case.

JackStouffer commented:
Did the builds really fail because of an extra space ಠ_ಠ ?

DmitryOlshansky commented:
> Did the builds really fail because of an extra space ಠ_ಠ ?

Yes it's on purpose to hunt down trailing spaces before they find their way into the repo

JackStouffer commented:
Is there anyway to include that check in the unit tests so I can find that out before I commit?

Also, it doesn't look like the auto tester is retesting my new commit as it still shows all of the old failures.

DmitryOlshansky commented:
> Is there anyway to include that check in the unit tests so I can find that out before I commit?

make -f posix.mak checkwhitespace

schveiguy commented:
> Also, it doesn't look like the auto tester is testing my new commit as it still shows all of the old failures.

Line 1646 still has trailing whitespace.

JackStouffer commented:
I'm making a shell script to run all of the whitespace checks when I run the unit tests. Any other checks that I should know about?

JackStouffer commented:
Rebased.

JackStouffer commented:
All tests now green! Unless someone has a suggestion algorithmically, I believe this is ready to be merged.

data-man commented:
>        int[string] r1_counts;
>        int[string] r2_counts;
>
>        foreach (item; r1) {
>           ++r1_counts[to!string(item)];
>        }

>        foreach (item; r2) {
>             ++r2_counts[to!string(item)];
>        }

The conversion to a string is not a good idea.

JackStouffer commented:
Yes, but is there another way to create keys for the AA?

quickfur commented:
AA's support arbitrary keys. (As long as said keys define `toHash` and `opEquals`, of course.)

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+    Returns:
+        $(D bool)
+*/
+enum AllocateGC {
```
Style nitpick: please put opening braces on their own line.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+    !isInfinite!Range1 &&
+    !isInfinite!Range1)
+{
+    if (allocate_gc == AllocateGC.yes) {
```
ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+        // If the two ranges aren't the same length, then they cannot
+        // possibly be permutations
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length")) {
+            if (r1.length != r2.length) {
```
ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+        int[string] r1_counts;
+        int[string] r2_counts;
+
+        foreach (item; r1) {
```
ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+            ++r1_counts[to!string(item)];
+        }
+
+        foreach (item; r2) {
```
ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+        }
+
+        return r1_counts == r2_counts;
+    } else {
```
Style nitpick: should be 3 separate lines:
````
}
else
{
````

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+        import std.algorithm.searching : count;
+
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length")) {
+            if (r1.length != r2.length) {
```
Opening brace on own line.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,114 @@ unittest
+            }
+        }
+
+        foreach (item; r1.save) {
```
Well, you get the idea. :-)

JackStouffer commented:
Fixed all issues and squashed commits.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+        // possibly be permutations
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length"))
+        {
+            if (r1.length != r2.length) {
```
Nitpick: this brace was overlooked, it seems.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+    }
+}
+
+bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2) if (
```
Nitpick: that trailing `if(` should be put on new line, merged with the following line?

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+}
+
+///
+@safe unittest {
```
Watch the opening brace.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+    assert(!isPermutation([1, 1], [1, 1, 1]));
+}
+
+@safe unittest {
```
Ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+            ++r2_counts[item];
+        }
+
+        return r1_counts == r2_counts;
```
This algorithm currently requires constructing two AA's. I wonder if we could reduce its allocation footprint by using `CommonType!(E1,E2)`, and using a single AA for both ranges. For the first range, increment the count; for the second, decrement the count. At any time when the count reaches 0, remove the element from the AA. Then at the end there's no need to do the expensive AA comparison, but just check whether it's empty.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+            {
+                return false;
+            }
+        }
```
I'm still not perfectly happy with this O(m*n) algorithm. Surely there's a way to do it better?

JackStouffer replied:
I can't see one that doesn't allocate with the GC.

quickfur replied:
In terms of overall complexity, I guess O(m*n) isn't really avoidable if we are not allowed to allocate.

However, some cases still bother me, e.g., if `r1==r2`, then this algorithm will do a lot of needless work.  Similarly, if `r1` and `r2` share an initial identical segment, then it will also do a lot of needless work.  Moreover, if there are a lot of duplicate elements, then we will call `count(item)` on the same `item` multiple times, which is again needless work.

Obviously, some of this redundant work will be unavoidable if we cannot allocate, but there are some simple cases that could use some optimization. E.g., if we keep track of the last item seen, then we can skip over runs of the same item in one of the ranges, e.g.:
````
ElementType!r1 lastItem;
while (!r1.empty)
{
    lastItem = r1.front;
    if (r1.save.count(lastItem) != r2.save.count(lastItem)
        return false;
    r1.popFront();

    // Skip over runs of the same item
    while (!r1.empty && r1.front == lastItem)
        r1.popFront();
}
````

quickfur replied:
Or, to be transient-safe:
````
R1 lastPos;
while (!r1.empty)
{
    lastPos = r1.save;
    if (r1.save.count(lastItem) != r2.save.count(lastItem))
        return false;

    while (!r1.empty && r1.front == lastPos.front)
        r1.popFront();
}
````

quickfur replied:
Hmm, just realized that the above two examples are wrong, we should iterate over a `save`d copy of `r1`, not `r1` itself, otherwise the counts will be off.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+        }
+    }
+
+    return sort(r1.save) == sort(r2.save);
```
Hold on here... How do we know that `r1` and `r2` are sortable? The sig constraints at least need to check that both ranges have swappable elements. Also, this needs to be documented that calling this function will mutate the ranges, which the user may or may not desire.

quickfur replied:
If the user doesn't want the ranges to be mutated, perhaps we could create index arrays and perform the sort on the index arrays instead of the actual ranges.

JackStouffer replied:
Unless I misunderstand what save does, the ranges are not mutated as `r1.save` is passed to `sort` and not `r1`.

quickfur replied:
`.save` only saves the current *position* of the range; it is not guaranteed to create a copy of the underlying elements.

Try it with arrays to see for yourself:
````
auto arr = [5,2,3,1,4];
auto saved = arr.save;
saved.sort();
writeln(arr);
````

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+    Returns:
+        $(D bool)
+*/
+enum AllocateGC
```
This enum needs to be placed before the ddoc comment, otherwise the generated ddocs will be attached to the wrong symbol.

quickfur replied:
Also, I wonder if this enum should be placed in a more general module, since it's pertinent not just to `std.algorithm.sorting`, but to any Phobos function that might want to give the user a choice of whether or not to allocate. In any case, it also needs its own ddoc header.

Alternatively, we could consider using `Flag!"allocateGC"` (from std.typecons) instead of inventing a new enum for what's simply an on/off switch.

JackStouffer replied:
I considered Flag!"allocateGC", but I don't see a benefit in using a template when a simple enum performs the exact same task almost in the same way.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,129 @@ unittest
+
+        // If the two ranges aren't the same length, then they cannot
+        // possibly be permutations
+        static if (hasMember!(Range1, "length") && hasMember!(Range2, "length"))
```
Shouldn't we just use `std.range.primitives.hasLength` here? Or is this the new way of checking for the existence of members?

quickfur commented:
I also wonder if a better name for this might be `isPermutationOf`, then it allows nicer-looking syntax with UFCS: `if (r1.isPermutationOf(r2)) { ... }`. This is mostly just a cosmetic suggestion, though.

JackStouffer commented:
`isPermutationOf` doesn't work as well for normal calling IMO. This 
```
isPermutationOf(r1, r2)
``` 
looks odd to me.

quickfur commented:
Fair enough, `isPermutation` is passable.

quickfur commented:
About the issue of sorting... I'm uneasy about this function sorting the incoming ranges behind the user's back. I don't like this kind of hidden side-effects. A better approach might be to use `static if` to check if the incoming ranges are `SortedRange`s. If so, just do a straight range comparison. If not, fallback to the hash algorithm (if GC) or the expensive algorithm (if no GC).  That way the user can explicitly sort the ranges for best results, or if not, it will still work, just a bit slower.

JackStouffer commented:
I can't test this now as I'm having problems with dmd at the moment, but what if I did something like this
```
sort(r1.map!(a => a.dup)) == sort(r2.map!(a => a.dup));
```
While slow, it would be a little faster than the non allocating case.

quickfur commented:
Are you sure the result of `map` is sortable? If it is, are you sure it doesn't also mutate the underlying array/range?

JackStouffer commented:
Fixed commented issues. I could not find a way to guarantee that the map function would not mutate the underlying range, so I just did the sorted range overload as recommended.

JackStouffer commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,197 @@ unittest
+    static assert(result2);
+
+    //enum result3 = isPermutation([1.1, 2.3, 3.5], [2.3, 3.5, 1.1], AllocateGC.yes);
+    //static assert(result3);
```
For some reason, this assert fails during CTFE but works normally. How does one debug functions during CTFE?

quickfur replied:
Allocating with GC may interact badly with CTFE, perhaps? I'm not 100% sure.

Debugging functions in CTFE is a pain, because you can't use a debugger and the compiler may not be in a state that corresponds with an actual runtime state, making it complicated to figure out what went wrong. You can try `pragma(msg)`, but that may or may not give you what you want.

JackStouffer replied:
This function fails this assert in CTFE: https://github.com/D-Programming-Language/dmd/blob/master/src/dinterpret.d#L3522

JackStouffer commented:
Ping?

quickfur commented:
Generally, I think the allocating variants of the functions are probably ready to merge. I'm not so sure about the non-allocating variants that introduce performance issues, especially those with O(n^2) complexity or more. Other reviewers please comment.  @jmdavis @yebblies @DmitryOlshansky @schveiguy

schveiguy commented:
According to the c++ documentation, it is non-allocating, and complexity is O(n^2) at the most (and O(n) at best). It does this via 2 heuristics I think:

1. skip the elements at the beginning where r1.front == r2.front, they are in the same order and don't need to be counted.
2. Skip counting elements which are not the first case of the element in r1. Since these have already been tested, no need to test them again.

2 is going to be a tad more difficult because we are using ranges and not iterators. It depends on being able to compare 2 ranges without having them compare the elements. In D, this is not consistent.

Heuristic 1 should be a no-brainer, and should be used before either version (allocating or not).

I also think that for non-length defining ranges, you should use walkLength after heuristic 1 to verify the lengths are the same, since that's O(n) and we can't do any better with this algorithm anyway.

quickfur commented:
Just because it's O(n), doesn't mean the constant factor isn't important. I think it's fine to just proceed without using `walkLength` when `length` is not defined. If during the iteration it's found that the lengths are unequal, return false. That way we prevent walking the ranges more than necessary.

quickfur commented:
As for (2), it's not really that hard, is it?  If the incoming ranges are sorted, all we need to do is to skip over runs of the same item. In fact, it could be as simple as `r1.group().equal(r2.group())` in this case.

If r1 is sorted but r2 isn't, we can still just iterate over `r1.group()` and count the head element of each group in r2 to make sure they match. As a bonus, if r2 is sorted but r1 isn't, just swap the two arguments and do the same thing.

If both are unsorted, then (2) doesn't apply anyway, so we proceed as the current code already does. You can still use `r1.group()` to skip over any consecutive runs of the same element, but it won't save very much because the same element might occur again later.

JackStouffer commented:
>  If the incoming ranges are sorted, all we need to do is to skip over runs of the same item.

Do you mean sorted but not via the same ordering? Because now if the two ranges are sorted then a simple equality is used on the two ranges.

schveiguy commented:
> If both are unsorted, then (2) doesn't apply anyway

yes it does.

For example:

```D
assert(isPermutation([1,2,3,1,2,3], [3,2,1,3,2,1]));
```

While iterating these two ranges, you only have to check the count of the `1`s once, the `2`s once, and the `3`s once. As you go through the second set, you can skip the count calls.

How this works in the C++ sample implementation is you first search through r1 for 1, and if you are at the first instance, then you check for the `1`s, otherwise, you already did it. Since you are going to search for the first one anyway to count, you are cutting short that branch.

> Just because it's O(n), doesn't mean the constant factor isn't important

The constant factor is 1, you only have to `walkLength` once :)

> As for (2), it's not really that hard, is it?

The issue I see is checking to see if you are at the first element of that instance. For C++ iterators, you say:

```C++
iter firstInstance = std::find(r1first, r1second, *rcur);
if(firstInstance == rcur) // at the first one
```

For D, `std.algorithm.find` returns a range. If, for instance, the range is an array, doing `firstInstance == rcur` would do a comparison of all the elements in those ranges! If we use `firstInstance is rcur`, then a reference range type may not compare correctly. This is why I said it's trickier. Because C++ defines iterator comparison to be identity, it has that ability to do the comparison.

quickfur commented:
@schveiguy Ah, I see what you mean now. Well, the solution is quite easy, actually. Keep an index of the current element in `r1`. At each element `e`, when computing the count of `e`, you have to scan `r1` anyway, so might as well scan it while also keeping track of the scanning index. If the first occurrence of `e` in the scanning loop has an index smaller than the current index, then you know that the element has been seen before, and you can abort the scan and move on to the next element.

This is sorta like the C++ iterator implementation, with indices in lieu of iterators.

schveiguy commented:
@quickfur yeah that would work.

JackStouffer commented:
Added recommended optimizations. @quickfur I'm not sure if your suggestion of keeping the index was meant to supersede your suggestions with the sorted range grouping, so I only implemented the former.

schveiguy commented:
I'm actually thinking the O(n) length check is required. For example:

```D
struct NoLength(R)
{
    R base;
    alias base this;
    @disable length();
    auto save() { return this;}
}

assert(isPermutation(NoLength!(int[])([1,2,3]),NoLength!(int[])([4,3,2,1])));
```

How do you fix this without checking the length of r2?

schveiguy commented:
I don't know if the sorted range optimization is valid either. Imagine a range sorted by part of the fields in the structures. They could be permutations, but not necessarily identical.

JackStouffer commented:
> I'm actually thinking the O(n) length check is required. For example:

That assert correctly fails in the current code. See line 1665

> I don't know if the sorted range optimization is valid either. Imagine two ranges sorted by part of the fields in the elements. They could be permutations, but not necessarily identical.

Ok, I'll just change it to the `std.algorithm.equal` function.

JackStouffer commented:
> Ok, I'll just change it to the std.algorithm.equal function.

Done

quickfur commented:
@schveiguy You're right, the length check is probably necessary, unless I can think of some clever way to achieve the same thing.

schveiguy commented:
> That assert correctly fails in the current code.

Ah good. I didn't see that before.

> Ok, I'll just change it to the std.algorithm.equal function.

No, that's not enough. For example:

```D
alias mytuple = Tuple!(int, int);

assert(isPermutation(assumeSorted!("a[0] < b[0]")([mytuple(1, 1), mytuple(1, 2)]), assumeSorted!("a[0] < b[0]")([mytuple(1,2), mytuple(1, 1)])));
```

JackStouffer commented:
> No, that's not enough. For example:

Ok, I added the ability to pass a custom predicate to the equal function and added your example in the unit tests.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+    $(D bool)
+*/
+
+alias AllocateGC = Flag!"AllocateGC";
```
The docs need to go after this, otherwise the alias will be documented and the function will not be.

schveiguy commented:
I think we should drop the sorted optimization. It's essentially `std.algorithm.equal`. In fact, you don't need the length check, since `equal` will do this already. It's very unlikely that having a special overload that doesn't quite work without extra template parameters is going to be useful in generic code, when you could just use `equal` when you know it applies.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+
+/**
+Returns true if all of the elements in $(D r1) are present in $(D r2). Otherwise,
+returns false.
```
Not quite a correct description. I think the conditions are that r1 and r2 have the same number of elements, and elements appear in r1 the same number of times as they do in r2.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
+{
```
Why no predicate for equality here?

JackStouffer replied:
I think the best solution is to document that equality for the allocating option is based upon the toHash method of the elements and cannot be changed. The non allocating option is there for situations where this is not good enough.

quickfur replied:
The allocation option simply *does not work* for any predicate other than true equality. So the predicate shouldn't even be specifiable in that case. A function should not take a parameter that it cannot implement. The `allocate_gc` parameter should be compile-time rather than runtime, and there should be two overloads:
````
bool isPermutation(AllocateGC allocGC, R1, R2)(R1 r1, R2 r2)
    if (allocateGC == AllocateGC.yes)
{
    ....
}

bool isPermutation(AllocateGC allocGC, alias pred = `a==b`, R1, R2)(R1 r1, R2 r2)
    if (allocateGC == AllocateGC.no)
{
    ...
}
````
Or something along these lines.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+
+If your parameters are forward ranges, then this function can allocate if the
+$(D AllocateGC.yes) flag is passed. This has the benefit of being much
+faster than the $(D AllocateGC.no) option.
```
Define "much faster" in terms of complexity:

With AllocateGC.yes, complexity is amortized O(n)
With AllocateGC.no, complexity is O(n) for identical permutations, and up to O(n^2) for different permutations.

schveiguy replied:
Maybe we should change "much faster" to "have better complexity". I can't see that allocating with the GC will be much faster for small ranges.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+
+Params:
+    r1 = Any finite forward range
+    r2 = Any finite forward range
```
Define allocate_gc parameter

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+can override the default $(D "a == b") predicate. See $(D binaryFun) for more
+details
+
+Non allocating forward range option: $(BIGOH r1.length * r2.length)
```
No, r1.length == r2.length is checked before anything, so it's O(n^2)

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+details
+
+Non allocating forward range option: $(BIGOH r1.length * r2.length)
+Allocating forward range option: $(BIGOH r1.length) + $(BIGOH r2.length)
```
Should note that an AA is used, which is amortized O(1), not exactly O(1).

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+        ulong r1_count;
+        ulong r2_count;
+
+        if (r1.save.walkLength != r2.save.walkLength)
```
A couple notes here:

1. You are checking length above for `.length` defining ranges. So no need to check in that case.
2. I realized you only need to check the minimum length here. So for instance, if r1.length i 5, and r2.length is 5000, walkLength will go through all 5000 elements, where it only really needs to get to 6 to realize it's not equal. I don't know if there's an easy way to do this without looping in parallel.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,228 @@ unittest
+
+            if (counts[item] == 0)
+            {
+                counts.remove(item);
```
Instead of this, do:

```D
if(--counts[item] < 0)
    return false;
```
If you check the lengths of r1 and r2 are the same, then you don't need to check the AA at the end, because if any element appears less in r2 than in r1, then some other element will appear more.

JackStouffer commented:
@schveiguy I have addressed all of your comments.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,196 @@ unittest
+
+/**
+Returns $(D true) if all of the elements in $(D r1) appear the same number of times
+in $(D r2). Otherwise, returns $(D false).
```
Nice, much better.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,196 @@ unittest
+            if (counts[item] < 0)
+            {
+                return false;
+            }
```
This only works if you verify the lengths are the same previously. You are only doing this in the case that r1.length and r2.length are defined. You need to do the walkLength, or use `enumerate` to verify the lengths are the same after looping through each range.

schveiguy replied:
Er.. don't need enumerate, just need some sort of counter to increment and decrement.

JackStouffer replied:
This doesn't seem to be true, as adding 

```
    ReferenceForwardRange!int r11 = new ReferenceForwardRange!int([1, 2, 3]);
    ReferenceForwardRange!int r12 = new ReferenceForwardRange!int([4, 2, 1, 3]);
    assert(!isPermutation(r11, r12, AllocateGC.yes));
```

To the unit tests works.

schveiguy replied:
Reverse the parameters, and it will not pass.

JackStouffer replied:
This passes

```
    auto r11 = new ReferenceForwardRange!int([4, 2, 1, 3]);
    auto r12 = new ReferenceForwardRange!int([1, 2, 3]);
    assert(!isPermutation(r11, r12, AllocateGC.yes));
```

schveiguy replied:
Are you sure? Working through it, I can't see how it would pass. Unless r11 and r12 have .length defined? I'm not familiar with `ReferenceForwardRange`.

After it goes through the first loop, I would expect counts to be [4:1, 2:1, 1:1, 3:1]. After going through the second loop, I'd expect counts to be [4:1, 2:0, 1:0, 3:0]. Where does it return false for this?

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,196 @@ unittest
+                {
+                    return false;
+                }
+            }
```
Almost :) Now if r1 is long and r2 isn't, then we have the same problem.

I was thinking this:
```D
auto r1s = r1.save;
auto r2s = r2.save;

while(!r1s.empty)
{
   if(r2s.empty)
      return false;
   r1s.popFront();
   r2s.popFront();
}
if(!r2s.empty)
    return false;
```

schveiguy replied:
And of course, no need to do this if the lengths have already been checked above (because both r1 and r2 define length)

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,196 @@ unittest
+                }
+            }
+
+            r2_count = r2.save.count(item);
```
change count to use the predicate

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,196 @@ unittest
+        {
+            --counts[item];
+
+            if (counts[item] < 0)
```
Do:
```D
if(--counts[item] < 0)
```
To avoid double lookup.

JackStouffer commented:
Sorry, I accidentally switched to a different PR and was running the wrong unit tests. Everything should be fixed now.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,203 @@ unittest
+        [mytuple(1, 1), mytuple(1, 2)],
+        [mytuple(1, 2), mytuple(1, 1)]
+    ));
+}
```
Can you fix this missing EOL?

schveiguy replied:
Note, this only showed up on github if you click the Files changed tab.

schveiguy commented:
One final nit -- in the case of specifying an alternate predicate and using `AllocateGC.yes`, the function will happily complete, but will likely be incorrect. Leaving it that way allows incorrect code to compile and run without error. I think we have a couple options. First, we could make 2 different functions, one that allocates (and doesn't take a predicate) and one that does not allocate. Second, we could assert when `AllocateGC.yes` is passed, but predicate is not "a == b".

@quickfur what are your thoughts?

JackStouffer commented:
>  First, we could make 2 different functions, one that allocates (and doesn't take a predicate) and one that does not allocate

The problem with this option is what if you want to not allocate for simple ranges you *have* to pass `"a == b"`. Secondly, the great thing about the current function signature is that if you want to allocate, it's really obvious that your doing it, where the above suggestion could lead people to forget that not passing a `pred` means they're allocating.

> Second, we could assert when AllocateGC.yes is passed, but predicate is not "a == b".

I think this is the correct option.

schveiguy commented:
> if you want to not allocate for simple ranges you have to pass "a == b"

The "a == b" is a template parameter, which is defaulting to "a == b". Why would you have to pass it?

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -738,7 +741,7 @@ template equal(alias pred = "a == b")
 
             //Lengths are the same, so we need to do an actual comparison
-            //Good news is we can sqeeze out a bit of performance by not checking if r2 is empty
+            //Good news is we can squeeze out a bit of performance by not checking if r2 is empty
```
Why is this change here? This should be a separate PR, to avoid mixing different kinds of changes into a single PR.

JackStouffer commented:
>The "a == b" is a template parameter, which is defaulting to "a == b". Why would you have to pass it?

Actually while trying to come up with an example, I realized that the real problem is that you would have to call the two functions different names as these two signatures
 
```
bool isPermutation(Range1, Range2)(Range1 r1, Range2 r2)
bool isPermutation(alias pred="a == b", Range1, Range2)(Range1 r1, Range2 r2)
```

both apply for this call

```
isPermutation([1, 2], [2, 1]);
```

So you would have to name it something like `isPermutationGC`, which bucks the style of all of Phobos.

quickfur commented:
Good point. A more pertinent question, though, is whether `pred` even makes any sense to begin with. What does it mean for two ranges to be permutations of each other via any predicate that isn't equality, anyway? That seems to defy the definition of "permutation". That would seem to suggest that we should just dispense with `pred`, and just pass `AllocateGC` as an optional parameter, either runtime or compile-time.

In the very rare case where the user actually needs to specify such a strange thing as non-equality-based permutation test, it should be easy enough to implement the predicate via a `map` call or two, that maps elements from either range into a common type that can be compared via equality, e.g.:
````
auto r1 = ...;
auto r2 = ...;
if (isPermutation(r1.map!(a => f(a)), r2.map!(a => g(a)))) { ... }
// where f(a) and g(a) map each element into some common type
````

JackStouffer commented:
I agree on your points about `pred`, but I want to hear @schveiguy's opinion before I change anything

schveiguy commented:
> whether pred even makes any sense to begin with

1. It's consistent with other functions in phobos (i.e. equal or count (as used here)).
2. map may be a viable solution if your predicate compares a single field, but what if it's multiple fields?

It's also one of those things that's "free", since you have to do some kind of comparison anyway.

As to the use cases, I'm not sure it matters too much. The fact is, this would leave isPermutation less powerful than nextPermutation, which uses a predicate.

schveiguy commented:
One benefit of having a GC specific function, the non-gc one can be `@nogc`.

I'm thinking more and more, the GC based one is just not viable here. It's a questionable optimization, tradeoff complexity for using the (generally slow) GC.

JackStouffer commented:
> One benefit of having a GC specific function, the non-gc one can be @nogc

If the allocate gc flag was a template parameter, then the code could be `static if`-ed to include or not include the AA right? Then the template would auto mark itself @nogc.

>It's a questionable optimization, tradeoff complexity for using the (generally slow) GC.

In my benchmarks, the GC allocating version tends to be about twice as fast.


```
void f1()
{
    immutable arr1 = [23,20,14,22,11,29,17,13,26,25,21,24,27,18,12,10,16,28,15,19];
    immutable arr2 = [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29];

    cast(void) isPermutation(arr1, arr2);
}

void f2()
{
    immutable arr1 = [23,20,14,22,11,29,17,13,26,25,21,24,27,18,12,10,16,28,15,19];
    immutable arr2 = [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29];

    cast(void) isPermutation(arr1, arr2, AllocateGC.yes);
}


void main() {
    import std.stdio;
    auto r = benchmark!(f1, f2)(10000);
    to!Duration(r[0]).writeln;
    to!Duration(r[1]).writeln;
}
```

```
$ ./test 
178 ms, 697 μs, and 9 hnsecs
86 ms, 572 μs, and 2 hnsecs
```

schveiguy commented:
Your test also is testing the runtime construction of the arrays, using the GC. Try constructing the arrays outside the test function.

The problem with these kinds of tests is you could incur a collection cycle at some point, which may not have anything to do with the test.

I'm not saying the test is giving an incorrect answer, I just think it's not correctly constructed.

schveiguy commented:
> Try constructing the arrays outside the test function.

or you could make the arrays static.

JackStouffer commented:
```
static immutable arr1 = [23,20,14,22,11,29,17,13,26,25,21,24,27,18,12,10,16,28,15,19];
static immutable arr2 = [10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25,26,27,28,29];

void f1()
{
    cast(void) isPermutation(arr1, arr2);
}

void f2()
{
    cast(void) isPermutation(arr1, arr2, AllocateGC.yes);
}


void main() {
    import std.stdio;
    auto r = benchmark!(f1, f2)(10000);
    to!Duration(r[0]).writeln;
    to!Duration(r[1]).writeln;
}
```

```
$ ./test
167 ms, 392 μs, and 8 hnsecs
71 ms and 238 μs
```

schveiguy commented:
Seems pretty definitive. I'm actually pleasantly surprised, the GC improvements have been very good in the last version.

Let's make the Allocate flag a template parameter, and I think it will work if you do this:
```D
bool isPermutation(AllocateGC allocate_gc, R1, R2)(R1 r1, R2 r2) if(allocate_gc == AllocateGC.yes) {...}
bool isPermutation(alias pred, R1, R2)(R1 r1, R2 r2) if(is(typeof(binaryFun!(pred)))) {...}
```

Adding in additional constraints for R1 and R2 to be proper ranges.

JackStouffer commented:
Added the new function signatures as requested. I tried to make the code as DRY as possible by creating the new function `checkLength`, but most of the code is subtly different between the two functions because of `predEquals`.

JackStouffer commented:
Actually, now that I think about it, would `checkLength` be a good public addition to `std.algorithm.comparison`? There is currently no function in `std` to check the lengths of two ranges like it does.

JackStouffer commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,289 @@ unittest
+
+    auto r7 = new ReferenceForwardRange!int([4, 2, 1, 3]);
+    auto r8 = new ReferenceForwardRange!int([1, 2, 3]);
+    assert(!isPermutation!(Flag!"AllocateGC".yes)(r7, r8));
```
For some reason, passing in `AllocateGC.yes` here gave an error.

schveiguy replied:
Still the case now that we fixed the signatures? I have a feeling this is an issue with the bug you reported.

JackStouffer replied:
> I have a feeling this is an issue with the bug you reported.

You're right. It works now.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,289 @@ unittest
+        {
+            return false;
+        }
+    }
```
If we're going through the trouble of making a special function to do this (especially if we want to make it public, and I think that's a good idea), we should make sure we don't walk the length of ranges with length. So if r1 has length, but r2 doesn't, or vice versa, only walk the length of one of them.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1568,289 @@ unittest
+    if (isForwardRange!Range1 &&
+        isForwardRange!Range2 &&
+        !isInfinite!Range1 &&
+        !isInfinite!Range2)
```
OK, so I don't think this works well. I think isPermutation(r1, r2) is ambiguous (I don't even know how this compiles), and we should utilize some better options here.

What I was talking about originally was something like this:
```D
bool isPermutation(AllocateGC allocate_gc, Range1, Range2)
(Range1 r1, Range2 r2)
    if (allocate_gc == AllocateGC.yes &&
        isForwardRange!Range1 &&
        isForwardRange!Range2 &&
        !isInfinite!Range1 &&
        !isInfinite!Range2)
```

There are other ways we can make this work, but I should be able to do:

```D
isPermutation(r1, r2); // calls non-allocating version
isPermutation!(AllocateGC.yes)(r1, r2); // calls allocating version
isPermutation!("a[0] == b[0]")(r1, r2); // calls non-allocating version with custom predicate
// isPermutation!(AllocateGC.yes, "a[0] == b[0]")(r1, r2); // error
// isPermutation!(AllocateGC.no, "a[0] == b[0]")(r1, r2); // not necessary to have this compile, but if it does, should do the right thing.
```

JackStouffer replied:
I'm a little confused by your suggestions here, because

1. You say `isPermutation(r1, r2)` is ambiguous but you list it in your examples of what should compile

2. All of your examples do as you ask in the current code.

schveiguy replied:
I meant that isPermutation(r1, r2) is ambiguous with your current code (it could match both template overloads). As I said, I have no idea why this compiles, I assume it's a compiler bug.

JackStouffer replied:
> I have no idea why this compiles, I assume it's a compiler bug

https://issues.dlang.org/show_bug.cgi?id=15063

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,385 @@ unittest
+If both ranges have a length member: $(BIGOH 1)
+If r1 doesn't have a length member: $(BIGOH r1.length)
+If r2 doesn't have a length member: $(BIGOH r2.length)
+If both ranges don't have a length member: $(BIGOH 2 * r1.length)
```
Nice addition!

For complexity, If both ranges have a length member, it's O(1). Otherwise, it's always O(min(r1.length, r2.length)). For big-O notation, you never include constants (e.g. 2 * r1.length).

JackStouffer commented:
Adressed concerns, and I added more tests.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,407 @@ unittest
+
+This function can allocate if the $(D AllocateGC.yes) flag is passed. This has
+the benefit of have better complexity than the $(D AllocateGC.no) option. However,
+this option is only available for ranges who's equality can be determined via each
```
who's => whose

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,407 @@ unittest
+this option is only available for ranges who's equality can be determined via each
+element's $(D toHash) method. If customized equality is needed, then the $(D pred)
+template parameter can be passed, and the function will automatically switch to
+the non-allocating algorithm. See $(D binaryFun) for more details on how to define
```
$(XREF functional,binaryFun) should provide a cross reference link instead of just a code-style element.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,407 @@ unittest
+    assert(!isSameLength(r11, r12));
+}
+
+alias AllocateGC = Flag!"AllocateGC";
```
Should document this, since it is the type used in the signature (appears that way in the docs too), and it is public.

just:

```D
/// For convenience
```
is probably sufficient.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,407 @@ unittest
+
+        return true;
+    }
+    else
```
Can get rid of this branch. assert(0) instead. This instantiation shouldn't happen unless allocate_gc is no.

JackStouffer commented:
Fixed all issues.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1571,367 @@ unittest
+    assert(!isSameLength("abcd", "xyz"));
+}
+
+// Test CTFE
```
Nice :)

JackStouffer commented:
Noticed some weird stuff going on with ddoc, so I fixed that.

schveiguy commented on std/algorithm/comparison.d:
```diff
@@ -36,6 +36,12 @@ $(T2 mismatch,
+)
+$(T2 isPermutation,
+        $(D isPermutation([1, 2], [2, 1])) returns $(D true).)
+)
```
Just noticed, these need to go before the parentheses. They aren't getting into the docs.

Also need to update the cheat sheet in std.algorithm itself (I think).

I'm satisfied with this addition after these fixes. Great work, thanks for sticking with it!

JackStouffer commented:
Seems to be good to go.

schveiguy commented on std/algorithm/package.d:
```diff
@@ -54,6 +54,8 @@ $(TR $(TDNW Comparison)
         $(SUBREF comparison, mismatch)
         $(SUBREF comparison, predSwitch)
+        $(SUBREF comparison, isSameLength)
+        $(SUBREF comparison, isPermutation)
```
One last nit, because I saw Walter complain about this elsewhere. Put these in alphabetical order in the list (and the cheat sheet for the local module).

schveiguy replied:
LGTM after that

JackStouffer commented:
Fixed.

schveiguy commented:
LGTM, ping @quickfur @DmitryOlshansky

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
 }
+
+/**
+Returns $(D true) if both ranges have the same length. This function is optimized
```
Seems redundant to have a sentence that says `Returns` and a `Returns:` section that states what is already obvious (the return type is explicitly named). What about moving this sentence after `Params:` and turn it into the `Returns:` section?

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+        if (r1.length != r2.length)
+        {
+            return false;
+        }
```
This is a really verbose way of doing something very simple. What about just:
````
return (r1.length == r2.length);
````
?

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+        if (length < r1.length)
+        {
+            return false;
+        }
```
ditto

JackStouffer replied:
`length < r1.length` can't be turned into a simple return statement, as `length > r1.length` is false as well.

schveiguy replied:
two options:

```D
return length >= r1.length;
return !(length < r1.length);
```

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+        if (length < r2.length)
+        {
+            return false;
+        }
```
ditto

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+    alias E1 = Unqual!(ElementType!Range1);
+    alias E2 = Unqual!(ElementType!Range2);
+
+    if(!isSameLength(r1.save, r2.save))
```
Nitpick: space after `if`.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+        {
+            break;
+        }
+    }
```
This is a very verbose way of doing something simple. What about:
````
while (!r1.empty && !r2.empty && predEquals(r1.front, r2.front))
{
    r1.popFront();
    r2.popFront();
}
````

JackStouffer replied:
I tried that, but if `r1.empty` is true, `r1.front` fails an assertion as all elements of the if statement are executed regardless of ordering.

schveiguy replied:
> I tried that, but if r1.empty is true, r1.front fails an assertion as all elements of the if statement are always executed regardless of ordering.

That seems incorrect. I will test and file a bug if this is the case. if statements should be short-circuit, evaluated left to right.

schveiguy replied:
I just tried it out, it works fine:

```D
import std.stdio;

int foo() { writeln("foo");return 0;}
int bar() { writeln("bar");return 1;}
int runboth(int x, int y) {writeln("runboth"); return 1;}
void main()
{
    if(foo() && bar() && runboth(foo(), bar()))
        writeln("here");
}
// output:
// foo
```

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+how to define $(D pred).
+
+Non allocating forward range option: $(BIGOH n^2)
+Non allocating forward range option with custom $(D pred): $(BIGOH n^2)
```
Nitpick: "non-allocating" should be hyphenated, not two separate words.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+        {
+            break;
+        }
+    }
```
This loop is also needlessly verbose. What about:
````
while (!r1.empty && !r2.empty && r1.front == r2.front)
{
    r1.popFront();
    r2.popFront();
}
````

JackStouffer replied:
See above comment.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+    alias E1 = Unqual!(ElementType!Range1);
+    alias E2 = Unqual!(ElementType!Range2);
+
+    if(!isSameLength(r1.save, r2.save))
```
Nitpick: space after `if`.

quickfur commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,367 @@ unittest
+    r2 = A finite forward range
+
+Returns:
+    $(D bool)
```
Probably better to move first sentence in the doc here, and in its place put a more general description like "determines if two ranges are permutations of each other".  The `Returns:` section should describe what's returned, not just repeat the return type when it's already obvious from the function signature.

quickfur commented:
Sorry for picking things apart again. :-)  Overall, it looks very good modulo said comments, and looks almost ready to merge. Thanks for the hard work in bringing this into good shape!

JackStouffer commented:
Fixed fix-able problems.

JackStouffer commented:
Actually fixed now.

schveiguy commented:
LGTM

schveiguy commented:
Auto-merge toggled on

JackStouffer commented:
Thanks!

MartinNowak commented on std/algorithm/comparison.d:
```diff
@@ -1565,3 +1569,340 @@ unittest
+}
+
+/// For convenience
+alias AllocateGC = Flag!"AllocateGC";
```
It's supposed to be `lias AllocateGC = Flag!"allocateGC";`, so that `Yes.allocateGC` has the correct casing.
See https://github.com/D-Programming-Language/phobos/pull/3717.
